feat: Built-in brute-force protection with IP blocking and user lockout

mrmm opened the pull request:
## Summary

This PR adds fail2ban-like login protection to Warpgate with:

- **IP-based rate limiting** - Automatically blocks IPs after configurable failed login attempts
- **Exponential backoff** - Each subsequent block has longer duration (30min → 1h → 2h → ... → 24h max)
- **User account lockout** - Locks accounts after repeated failures (optional auto-unlock)
- **Cross-protocol protection** - Failed attempts across SSH, HTTP, MySQL, PostgreSQL all count together
- **Admin management** - View status, unblock IPs, unlock users via Admin UI or API
- **In-memory caching** - Fast lookups with periodic database sync

## Changes since last review

- **Rebased onto upstream/main** (resolved 11 conflict files from 21 new upstream migrations/features)
- **Config moved into `parameters` DB table** (per Eugeny's review request): all login protection settings are now stored in the `parameters` table via migration `m00047`, editable at runtime via the existing parameters admin API endpoint — no `warpgate.yaml` config needed
- **Migration renumbered**: `m00025_login_protection` → `m00046_login_protection` to avoid conflict with upstream's m00025–m00045
- **chrono → time crate** migration across all login protection entities, cache, and service

### Configuration (now via `PUT /@warpgate/admin/api/parameters`)

All login protection settings are part of the existing parameters endpoint:

```json
{
  "login_protection_enabled": true,
  "login_protection_retention_days": 30,
  "lp_ip_max_attempts": 5,
  "lp_ip_time_window_minutes": 15,
  "lp_ip_base_block_duration_minutes": 30,
  "lp_ip_block_duration_multiplier": 2.0,
  "lp_ip_max_block_duration_hours": 24,
  "lp_ip_cooldown_reset_hours": 24,
  "lp_ip_blocked_message": null,
  "lp_user_max_attempts": 10,
  "lp_user_time_window_minutes": 60,
  "lp_user_auto_unlock": false,
  "lp_user_lockout_duration_minutes": 60,
  "lp_user_locked_message": null
}
```

The old `login_protection:` YAML config key is deprecated (skipped on serialization, ignored on read).

### Admin API Endpoints (all under `/@warpgate/admin/api`)

- `GET /parameters` / `PUT /parameters` - Read/write login protection settings
- `GET /login-protection/status` - Security overview (blocked IP count, locked user count)
- `GET /login-protection/blocked-ips` - List blocked IPs
- `DELETE /login-protection/blocked-ips/:ip` - Unblock an IP
- `GET /login-protection/locked-users` - List locked users
- `DELETE /login-protection/locked-users/:username` - Unlock a user

## Test plan

- [x] Docker Compose local setup - containers start correctly
- [x] IP blocking test - after 3 failed attempts, returns `{"state":"IpBlocked"}`
- [x] User lockout test - account locks after threshold
- [x] Integration tests - pytest tests pass
- [x] Build - `cargo build` succeeds (all changed crates, nightly-2025-10-21)
- [x] Admin UI - Login Protection page shows stats and lists
- [x] Rebase onto upstream/main - clean merge

mrmm committed:
feat(login-protection): add configuration and error types

Add LoginProtectionConfig, IpRateLimitConfig, and UserLockoutConfig
structs with sensible defaults. Add IpBlocked and UserLocked error
variants for authentication flow.

mrmm committed:
feat(login-protection): add database entities and migration

Add FailedLoginAttempt, IpBlock, and UserLockout entities.
Create m00025_login_protection migration with composite indexes
for efficient queries on IP/timestamp and username/timestamp.

mrmm committed:
feat(login-protection): implement core service with cache

Add LoginProtectionService with in-memory caching for fast lookups.
Implement IP blocking with exponential backoff, user lockout,
and background cleanup task for expired records.

mrmm committed:
docs: add login protection feature to README

- Add bullet point about built-in brute-force protection
- Add Documentation section with links to external docs
- Add comparison table row for brute-force protection

mrmm committed:
feat(protocols): integrate login protection across all protocols

- SSH: Check IP blocks before auth, record failed attempts
- HTTP: Check IP blocks and user lockouts in auth flow
- MySQL: Add login protection checks to session authentication
- PostgreSQL: Add login protection checks to session authentication

All protocols now share the same login protection service for
consistent brute-force protection across the entire gateway.

mrmm committed:
feat(admin-api): add login protection management endpoints

- GET /login-protection/status - Security status overview
- GET /login-protection/blocked-ips - List blocked IPs
- DELETE /login-protection/blocked-ips/:ip - Unblock an IP
- GET /login-protection/locked-users - List locked users
- DELETE /login-protection/locked-users/:username - Unlock a user

All endpoints require admin authentication.

mrmm committed:
chore: regenerate OpenAPI schemas for login protection endpoints

Updates admin and gateway OpenAPI schemas with new login protection
types and endpoints.

mrmm committed:
feat(ui): add Login Protection admin page

- New LoginProtection.svelte page with:
  - Stats cards (blocked IPs, locked users, failed attempts)
  - Blocked IPs list with unblock action
  - Locked users list with unlock action
- Add navigation link in Config.svelte

mrmm committed:
test: add login protection local testing setup

- Docker Compose setup with warpgate and SSH target
- Test configuration with aggressive settings for quick testing
- Test scripts for IP blocking and user lockout scenarios
- README with testing instructions

mrmm committed:
test: add login protection integration tests

Comprehensive pytest tests covering:
- IP blocking after failed attempts threshold
- Exponential backoff for repeat offenders
- User lockout functionality
- Admin API endpoints (status, unblock, unlock)
- Cross-protocol attempt counting

mrmm committed:
chore: regenerate config-schema.json for login protection

Adds LoginProtectionConfig, IpRateLimitConfig, and UserLockoutConfig
definitions to the JSON schema.

LarsSven commented:
@Eugeny what is the reason that this never got merged? We are still very interested in having some better brute-force protections

Eugeny commented:
It was closed by the author shortly after opening.
I haven't had time to work on an alternative yet but I'm 100% open to getting this finished or another PR

mrmm commented:
@Eugeny If the way it is implement match the expected structure I can refurbish it and rebase on the mast main commit ! and we can avoid having to redo all the work ofc! 

Just Let me know, I can look into it this weekend 🙏

LarsSven commented:
For us it would be very helpful if you could!

Eugeny commented:
Absolutely! I went through it and it's already 90% there. If you could rebase it and move the new config parameters into the Parameters model (I'm trying to avoid new config-only params), I can do another pass and adjust the remaining minor things.

LarsSven commented:
@mrmm have you perhaps had a chance to look at reopening it? Otherwise I could look into it but I would only have time a few months from now :/

mrmm committed:
chore: merge upstream/main into mm/login-protection-rebase

- Renamed migration m00025 → m00046 (upstream added 21 new migrations)
- Kept all login protection conflict resolutions in error.rs, services.rs,
  db-entities, db-migrations, auth.rs, mysql/postgres/ssh session handlers
- JSON schemas taken from upstream (will regenerate after T-002)

mrmm committed:
feat: move LoginProtectionConfig params into Parameters DB model

- Rename migration m00025 → m00046 to avoid upstream conflict
- Add m00047: ALTER TABLE parameters to add 14 login protection columns
- Add 14 fields to Parameters entity with defaults in Entity::get()
- Extend ParameterValues/ParameterUpdate in admin API for login protection
- LoginProtectionService::new() now takes Parameters::Model from DB
- services.rs fetches Parameters from DB; passes to LoginProtectionService
- Mark WarpgateConfigStore::login_protection as deprecated/skip_serializing
- Convert chrono→time in all login protection entities, cache, service
- Fix auth.rs: get_client_ip import path + correct call signature
- Regenerate config-schema.json and openapi schemas

mrmm committed:
fix: split SQLite ALTER TABLE into individual calls; use AuthenticatedRequestContext in login protection API

mrmm committed:
feat(web): add login protection section to admin Parameters page

Surfaces all 14 login_protection_* / lp_ip_* / lp_user_* DB parameters
in the admin UI: master enable switch, retention days, IP rate-limit
controls (max attempts, time window, base/multiplier/max block duration,
cooldown reset, custom blocked message) and user lockout controls
(max attempts, time window, auto-unlock toggle, lockout duration,
custom locked message). Validation requires positive integers; empty
optional message fields are omitted from PUT so the stored value is
preserved.

Linked to existing /@warpgate/admin/#/config/login-protection page for
blocked-IP / locked-user management.

mrmm committed:
fix(web): compact login protection layout into responsive grid

Was 14 stacked full-width FormGroups + 4 InfoBox cards (~2 screens).
Now uses Bootstrap row/col-md-3 grid: header + enable switch share one
line, retention on its own narrow input, IP rate-limiting fields wrap
4-per-row on md (max attempts | window | base block | multiplier on
row 1; max block | cooldown | message on row 2), user lockout fields
3-4 per row with auto-unlock toggle inline. Shorter labels, condensed
helper text, smaller subsection headers (h6 muted uppercase). Reduces
vertical footprint by ~70% on a 1080p screen.

mrmm committed:
fix(web): rewrite login protection UI as sentence-style settings

Previous compact grid still didn't work — col-md-3 floating-label inputs
inside a 768px container left ~150px columns with truncated labels and
no rhythm. Total height was ~6× the rest of the page.

New layout is one settings card with 5 rows of natural-language settings:

  IP rate-limit  Block after [5] failed attempts in [15] min for [30]
                 min, ×[2.0] on repeat, capped at [24] h, reset after
                 [24] h idle.
  User lockout   Lock account after [10] failed attempts in [60] min.
                 (•) Auto-unlock [after [60] min if on].
  Retention      Keep blocks, lockouts, and attempt history for [30]
                 days.
  Blocked-IP message  [text input full-width]
  Locked-user message [text input full-width]

Number inputs are 4rem wide (size matches actual content: small
positive integers), inline with descriptive prose so each setting reads
left-to-right like a sentence. Reduces section height from ~1100px to
~210px (~80% shorter) while making each value's meaning obvious without
a label-per-input.

Scoped <style> block. No new deps. Light/dark friendly via bootstrap
CSS vars.

mrmm committed:
fix(web): use SPA link for login protection page

Was a hard <a href> with /@warpgate/admin/ prefix that didn't resolve.
Use svelte-spa-router's use:link directive with bare /config/login-protection,
matching the Targets/AccessRoles pattern. Stays within the SPA, no reload.

mrmm committed:
fix(web): polish login protection UI — a11y, theming, mobile

After running design audit on the login protection section, addressed:

a11y (was P0):
- Add aria-label to every numeric input describing the policy field it
  controls (max attempts, time window, block duration, multiplier, cooldown,
  retention, lockout). Screen readers now announce purpose instead of
  'edit, blank'.
- Wrap each sentence in role='group' with aria-label for the policy as a
  whole.
- Convert the two visible labels next to the message inputs from <span>
  to <label for=...> with matching ids.
- Mark decorative section labels (IP rate-limit / User lockout / Retention)
  aria-hidden since the visible text is redundant with role='group' label.

input consistency:
- Drop custom border / padding / focus shadow on .lp-num; inherit
  Bootstrap .form-control-sm. Numeric inputs now match the message inputs
  and the theme's focus ring.
- Add max attributes (sane upper bounds: 1000 attempts, 1440 min = 1 day,
  720 h = 30 days, 3650 days = 10 years, 10080 min = 1 week, multiplier 10)
  so browsers can block obviously-bad input.
- On invalid value (NaN, < min): revert the displayed value to the last
  saved value instead of leaving the bad value visually in the field.
- :invalid border-color uses --bs-danger so out-of-range values are red.

responsive:
- Below 576px, stack the policy label above the sentence and tighten
  line-height; sentence runs full width instead of squeezing alongside
  a 9rem label on a phone.

clarity:
- Multiplier input gets a title= explaining 'each repeat block multiplies
  the previous duration by this factor (e.g. 2.0 doubles each time)'.

mrmm committed:
test: add AdminApiTestCase entries for login protection API ops

LarsSven commented:
@mrmm we are still super interested in this feature and are looking forward to it. Do you know what the status of the PR is?

mrmm commented:
@LarsSven cc @Eugeny 
So sorry for the delay, I Will try to get this out of the DARFT this week will ping you for review and approval 🙏

mrmm committed:
chore: merge upstream/main into mm/login-protection-rebase (renumber migrations m00053/m00054)

mrmm committed:
chore: add tracing::warn import; regenerate config + OpenAPI schemas after upstream/main merge

mrmm committed:
fix(web): login protection UI — UX, a11y, and reliability improvements

- Parameters.svelte: add error handling to update() (was silent on failure)
- Parameters.svelte: show dismissible Alert on save error
- Parameters.svelte: debounce numeric field saves (400ms, coalesces rapid edits)
- Parameters.svelte: show LP settings dimmed when disabled (pre-configure before enabling)
- Parameters.svelte: move Auto-unlock toggle to its own row (out of inline sentence)
- Parameters.svelte: clarify IP sentence ('within X min → block for Y min initially')
- Parameters.svelte: multiplier step 0.1→0.5, min=1.0 (browser spinner usable)
- Parameters.svelte: capitalise 'Login Protection' link text
- LoginProtection.svelte: add loading state with DelayedSpinner
- LoginProtection.svelte: add error display with Retry button
- LoginProtection.svelte: auto-refresh every 30s + manual Refresh button
- LoginProtection.svelte: Unblock/Unlock require confirmation dialog
- LoginProtection.svelte: disable action buttons while request in-flight
- LoginProtection.svelte: replace h2 count labels with semantic section-header
- LoginProtection.svelte: capitalise h1 'Login Protection'
- LoginProtection.svelte: remove spacer div, use mb-3 on list-group

mrmm committed:
feat(test): local-testing login-protection stack (start/seed/stop scripts)

- docker-compose.yml: ssh-target only (linuxserver/openssh-server:2223)
- start.sh: unattended-setup + warpgate run --enable-admin-token + wait
- seed.sh: creates role/user/password-cred/SSH-target via admin API
- stop.sh: kills warpgate process + docker compose down
- .gitignore: exclude data/, .warpgate.pid, warpgate.log
- drop old m00046/m00047 migration stubs (renamed to m00053/m00054 in merge commit)

mrmm committed:
fix(web): LP section — all impeccable-critique findings

- Add $derived lpCapWarning: initial block > max cap → inline Alert
- Toggle label: 'Enable brute-force protection' (verb-first, replaces noun phrase)
- Add InfoBox explaining feature (dimming-preserves-settings note)
- Remove aria-disabled from div container (individual disabled attrs sufficient)
- Background: var(--bs-tertiary-bg) + border (was hardcoded rgba hack)
- IP sentence: 'Initial block: X min × Y per repeat, max Z h. Repeat count resets
  after N h with no failed attempts.' (unambiguous, clean label separation)
- Add backoff-model InfoBox inside policy block (multiplier × prev duration,
  repeat count resets on clean cooldown not on block expiry)
- User lockout: auto-unlock row moves to own indented row, no empty spacer span
- Auto-unlock text responds to state: shows duration when on, 'off — manual
  unlock required' when off (no confusing parenthetical)
- Retention: 'Keep failed attempt and lockout records' (not 'blocks, lockouts')
- Messages: switch from lp-row to FormGroup floating (page-consistent pattern)
- Messages: oninput + scheduleUpdate (was onchange + update — saves on blur only)
- Add InfoBox on messages explaining where they appear + format constraints
- Footer link moved OUTSIDE lp-block (was inside — unclickable when LP disabled)
- Link text: 'Login protection' (sentence case, matches Section title)
- CSS: :global(.lp-num:invalid) with box-shadow (was scoped, unreliable)
- CSS: remove .lp-row-msg, .lp-row-auto-unlock (dead)
- CSS: lp-row-indent replaces empty-span spacer hack
- CSS: lp-auto-unlock-off for muted off-state label

mrmm committed:
fix(web): LP — split IP row into 2 lines, transparent block bg for dark mode

- IP rate-limit split into two rows:
  row 1: 'Block after [N] attempts within [N] min.' (trigger)
  row 2 (indent): 'Block for [N] min initially, × [N] per repeat, max [N] h. Resets after [N] h clean.' (caps)
  Prevents mid-sentence flex-wrap at 'min. Initial block:' and orphaned 'h' tokens
- lp-block background: var(--bs-tertiary-bg) → transparent
  --bs-tertiary-bg resolves to near-white in Warpgate dark theme; transparent
  keeps the block theme-agnostic; border alone provides sufficient grouping

mrmm committed:
fix(web): LP section — full restructure to FormGroup grid, drop bespoke CSS

Replace sentence-style flex layout with Bootstrap row/col grid + FormGroup floating:
- IP rate-limit: 2×2 grid (trigger: 2 cols; block/caps: 4 cols)
  'Max failures before IP block' | 'Failure window (minutes)'
  'Initial block (min)' | 'Backoff multiplier' | 'Max block (hours)' | 'Cooldown reset (hours)'
- User lockout: 2-col grid + auto-unlock toggle + conditional duration FormGroup
- Retention: single full-width FormGroup
- Custom messages: 2 full-width FormGroup (already done)
- All subsections under lp-group-title headings (uppercase caption style)
- Dimming: single .lp-disabled wrapper div (opacity + pointer-events)
  replaces per-input disabled pattern for visual state; each input keeps
  disabled attr for a11y
- CSS: drop 90-line bespoke system (.lp-block, .lp-row, .lp-sentence,
  .lp-label, .lp-num, .lp-row-indent, etc.) — keep only .lp-disabled,
  .lp-group-title, .lp-foot (3 classes, 20 lines)

mrmm committed:
fix(core): LP config reads Parameters from DB per-call, no startup snapshot

Match the same hot-reload pattern used by every other warpgate parameter
(ssh_client_auth_*, ticket_*, record_scp, minimize_password_login etc.)
which all call Parameters::Entity::get(&db) directly at the point of use.

Changes:
- LoginProtectionService: drop config field entirely; add read_config()
  private helper that calls Parameters::Entity::get() on each method
  invocation — same as ssh/ticket/scp parameter reads
- new() now takes only db: Arc<Mutex<DatabaseConnection>> (no params arg);
  reads config once to warm the cache, then discards it
- All public methods (check_ip_blocked, check_user_locked, record_failed_attempt,
  cleanup_expired, clear_failed_attempts, list_blocked_ips, list_locked_users)
  now read config from DB at their start — zero restart needed after UI change
- create_or_update_ip_block / create_user_lockout: made associated fns
  (no self receiver) taking config as param; removes redundant self.config access
- services.rs: remove db_params pre-fetch + is_enabled() guard; always start
  cleanup task (cleanup_expired() itself reads enabled from DB each run)
- services.rs: drop unused Parameters import
- parameters.rs: remove reload_config hook (no longer needed)

mrmm committed:
test: expand login protection test suite — IP block, user lockout, hot-reload

Previous coverage: 5 smoke/endpoint tests only (no actual block/lockout assertions).

New tests:
- test_ip_blocking_triggers_and_blocks_correct_password
  Fire ip_max bad attempts → verify IpBlocked state (incl. correct password)
  then admin unblock → verify correct password succeeds
- test_user_lockout_triggers_and_blocks_correct_password
  Fire user_max bad attempts → verify UserLocked state (incl. correct password)
  then admin unlock → verify correct password succeeds
- test_user_lockout_auto_unlock
  Trigger lockout with auto_unlock=True → verify locked state
- test_config_hot_reload_without_restart  ← key regression test for fix
  1. Start with user_max=5
  2. Make 3 bad attempts (below threshold)
  3. Lower user_max to 2 via admin API — NO restart
  4. One more attempt → must lock (total 4, new threshold 2)
  Fails if LP still reads startup snapshot instead of DB per-call
- test_config_hot_reload_raising_threshold
  Raise threshold mid-session → 4 attempts must NOT lock at old threshold=3

Also: _lp_wg() helper now sets thresholds via admin API (Parameters DB)
instead of config_patch (YAML) — LP config is DB-only.

mrmm committed:
fix(admin-api): LP endpoint security, IPv6 routing, test accuracy

Security:
- Add require_admin_permission() to all 5 LP handlers (was missing entirely)
- GET endpoints: require_admin_permission(&ctx, None)  — any authenticated admin
- DELETE endpoints (unblock/unlock): require_admin_permission(&ctx, Some(ConfigEdit))
  — matches the permission level of PUT /parameters for LP settings

IPv6 routing fix:
- DELETE /login-protection/blocked-ips/:ip → DELETE /login-protection/blocked-ips?ip=
  Path segment 'ip' causes ambiguity with IPv6 addresses (e.g. ::1 contains colons).
  Query parameter avoids any URL-parser edge case and is standard for filter/select
  actions that are not hierarchical resource identifiers.
- unlock_user stays as DELETE /locked-users/:username (usernames are safe in paths)
- OpenAPI schema regenerated to reflect the new query-param signature
- TypeScript admin client regenerated (gitignored, rebuilt at test time)

Tests:
- Fix credential endpoint: POST /users/{id}/credentials → /credentials/passwords
  with body {password: ...} (was silently returning 404, users had no passwords)
- seed.sh: same fix for locktest users
- test_config_hot_reload_without_restart: add extra attempt after threshold fires
  (lockout is created during attempt N; check_user_locked sees it on attempt N+1)
- test_config_hot_reload_raising_threshold: improved assertion messages
- All 4 live test sections pass: permissions, IPv6 unblock, lockout+unlock, hot-reload

mrmm committed:
chore: add openapitools.json (pinned generator version 7.23.0)

mrmm committed:
fix(ci): regenerate config-schema.json + fix AdminApiTestCase permissions

config-schema.json:
  Regenerated — upstream added SsoReturnUrlDomainPreference enum type and
  return_url_domain field to SsoProvider in the 25 commits merged from main.
  Local generation from the merge commit caught these; committed now so the
  'Config schema check' CI job passes (job runs 'just config-schema' and diffs).

tests/test_api.py:
  unblock_ip  permission None → 'config_edit'
  unlock_user permission None → 'config_edit'
  Both LP mutation endpoints now require AdminPermission::ConfigEdit (same as
  PUT /parameters). The AdminApiTestCase framework tests that endpoints return
  403 when the permission is absent, so these entries need the correct permission
  declared to pass the negative-case check.

Note: check-schema-compatibility will still warn about the new IpBlocked and
UserLocked enum values added to the gateway auth state response — this is
intentional (the PR adds these response states) and is a breaking-change review
signal for the maintainer, not a bug.

mrmm committed:
fix(admin-api): make ParameterUpdate.allow_own_credential_management optional

Previously bool (required), which caused pydantic ValidationError in the
Python SDK when calling sdk.ParameterUpdate() with only LP-specific fields
(e.g. from test helpers that do partial updates).

All other boolean fields in ParameterUpdate are Option<bool> — this was an
oversight. Partial PUT /parameters must work for any subset of fields.

Also regenerate openapi-schema.json (allow_own_credential_management is now
nullable in the schema) and config-schema.json.

mrmm commented:
Hi @Eugeny 👋

Picking this back up — addressed everything from your May review:

- **Config params in Parameters model** (migrations m00053/m00054, no new YAML key)
- **Hot-reload** — reads `Parameters::Entity::get(&db)` per call, same pattern as `ssh_client_auth_*`. No restart needed after a settings change.
- **10 integration tests** including hot-reload regression tests.
- Misc: IPv6-safe unblock (`?ip=` query param), FormGroup UI (dark-mode, no bespoke CSS), `ParameterUpdate` made fully partial.

One CI note: `check-schema-compatibility` warns on the new `IpBlocked`/`UserLocked` enum values in the gateway auth state — intentional, but happy to discuss if you prefer a different approach.

cc @LarsSven

Eugeny committed:
cleanup, fixes, admin exemption setting

Eugeny committed:
store parameters as seconds, UI tweaks

Eugeny committed:
test fixes, better spot to record failure

Eugeny committed:
prefer post

Eugeny commented:
Thank you! I've done a bit of clean up, removed dead code and added an extra override to prevent admin lockout. Tweaked the UI as well. I'll test it on my internal deployment for a bit that's exposed to the internet, taking bets on what the numbers are going to be :D

mrmm commented:
@Eugeny Thanks for reviewing this! Can you please let me know also of the number and curious 🙏
